src/java/org/apache/cassandra/index/sai/disk/v1/WidePrimaryKeyMap.java

```diff
@@ -0,0 +1,177 @@
+                                         partitionKeyReader.openCursor(),
+                                         clusteringKeyReader.openCursor(),
+                                         partitioner,
+                                         primaryKeyFactory,
+                                         clusteringComparator);
+        }
+
+        @Override
+        public void close()
+        {
+            super.close();
+            FileUtils.closeQuietly(Arrays.asList(clustingingKeyBlocksFile, clusteringKeyBlockOffsetsFile));
+        }
+    }
+
+    private final ClusteringComparator clusteringComparator;
+    private final SortedTermsReader.Cursor clusteringKeyCursor;
+
+    private WidePrimaryKeyMap(LongArray tokenArray,
+                              LongArray partitionArray,
+                              SortedTermsReader.Cursor partitionKeyCursor,
+                              SortedTermsReader.Cursor clusteringKeyCursor,
+                              IPartitioner partitioner,
+                              PrimaryKey.Factory primaryKeyFactory,
+                              ClusteringComparator clusteringComparator)
+    {
+        super(tokenArray, partitionArray, partitionKeyCursor, partitioner, primaryKeyFactory);
+
+        this.clusteringComparator = clusteringComparator;
+        this.clusteringKeyCursor = clusteringKeyCursor;
+    }
+
+    @Override
+    public long rowIdFromPrimaryKey(PrimaryKey primaryKey)
+    {
+        long rowId = tokenArray.indexOf(primaryKey.token().getLongValue());
+
+        // If the key only has a token (initial range skip in the query), the token is out of range,
+        // or we have skipped a token, return the rowId from the token array.
+        if (primaryKey.isTokenOnly() || rowId < 0 || tokenArray.get(rowId) != primaryKey.token().getLongValue())
+            return rowId;
+
+        // Search the sorted terms for the key in the same partition
+        return clusteringKeyCursor.partitionedSeekToTerm(clusteringComparator.asByteComparable(primaryKey.clustering()), rowId, startOfNextPartition(rowId));
+    }
+
+    @Override
+    public void close()
+    {
+        super.close();
+        FileUtils.closeQuietly(clusteringKeyCursor);
+    }
+
+    @Override
+    protected PrimaryKey supplier(long sstableRowId)
+    {
+        return primaryKeyFactory.create(readPartitionKey(sstableRowId), readClusteringKey(sstableRowId));
+    }
+
+    private Clustering<?> readClusteringKey(long sstableRowId)
+    {
+        ByteSource.Peekable peekable = ByteSource.peekable(clusteringKeyCursor.seekForwardToPointId(sstableRowId)
+                                                                              .asComparableBytes(ByteComparable.Version.OSS50));
+
+        Clustering<?> clustering = clusteringComparator.clusteringFromByteComparable(ByteBufferAccessor.instance, v -> peekable);
+
+        if (clustering == null)
+            clustering = Clustering.EMPTY;
```

Comment:
nit: Apologies...I keep forgetting what tests hits this code path (empty clustering) every time I review and run code coverage :D

Reply:
I'm not sure about a specific test for it, but this will disappear with CASSANDRA-18734. The only time we have a null clustering is when we create a partition-only key for the segment metadata.

I'd like to defer this to 18734.

Reply:
Cool, no worries if it's going away anyway...